## Medical Insurence Predictor


#### Life cycle of Machine learning Project

- Understanding the Problem Statement
- Data Collection
- Data Checks to perform
- Exploratory data analysis
- Data Pre-Processing
- Model Training
- Choose best model

### 1) Problem statement

- The project is about male and female population of smoker in an age group , does they have children or not add on which region they come from  , Calculating BMI (BODY MASS INDEX), and how much insurence they charge

### 2) Data Collection
- Dataset Source - https://www.kaggle.com/datasets/rahulvyasm/medical-insurance-cost-prediction/data
- The data consists of  7column and 2700 rows.

### 2.1 Import Data and Required Packages
####  Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv('data\medical_insurance.csv')

In [4]:
df

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520
...,...,...,...,...,...,...,...
2767,47,female,45.320,1,no,southeast,8569.86180
2768,21,female,34.600,0,no,southwest,2020.17700
2769,19,male,26.030,1,yes,northwest,16450.89470
2770,23,male,18.715,0,no,northwest,21595.38229


In [6]:
df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


In [8]:
df.shape

(2772, 7)

### 2.2 Dataset information

- Sex : Male,Female
- Age 
- Bmi :Body mass count
- Smoker : YES / NO
- Region : Northeast, Northwest, Southeast, Southwest
- Charges: Insurance Charge




### 3. Data Checks to perform

- Check Missing values
- Check Duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in the different categorical column

### 3.1 Check Missing values

In [9]:
df.isna().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

#### There are no missing values in the data set

### 3.3 Check data types

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2772 entries, 0 to 2771
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       2772 non-null   int64  
 1   sex       2772 non-null   object 
 2   bmi       2772 non-null   float64
 3   children  2772 non-null   int64  
 4   smoker    2772 non-null   object 
 5   region    2772 non-null   object 
 6   charges   2772 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 151.7+ KB


### 3.4 Checking the number of unique values of each column

In [11]:
df.nunique()

age           47
sex            2
bmi          548
children       6
smoker         2
region         4
charges     1337
dtype: int64

### 3.5 Check statistics of data set

In [12]:
df.describe()

,age,bmi,children,charges
count,2772.000000,2772.000000,2772.000000,2772.000000
mean,39.109668,30.701349,1.101732,13261.369959
std,14.081459,6.129449,1.214806,12151.768945
min,18.000000,15.960000,0.000000,1121.873900
25%,26.000000,26.220000,0.000000,4687.797000
50%,39.000000,30.447500,1.000000,9333.014350
75%,51.000000,34.770000,2.000000,16577.779500
max,64.000000,53.130000,5.000000,63770.428010


- Age Distribution: The dataset includes adults aged 18 to 64, with a median age of 39, suggesting a balanced age range.

- BMI Range: BMI values range from 15.96 to 53.13, with a mean of 30.70, indicating the presence of individuals with varying body mass indexes.

- Children: Most individuals have 0 to 2 children, with a few having up to 5.

- Charges: Medical charges show significant variation, with a median of around $9333.01 and a wide range, highlighting differing healthcare costs across individuals.


### 3.7 Exploring Data

In [13]:
df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


In [18]:

print("Categories in 'sex' variable:  ",end=" ")
print(df['sex'].unique())


print("Categories in 'smoker' variable:  ",end=" ")
print(df['smoker'].unique())


print("Categories in 'region' variable:  ",end=" ")
print(df['region'].unique())

Categories in 'sex' variable:   ['female' 'male']
Categories in 'smoker' variable:   ['yes' 'no']
Categories in 'region' variable:   ['southwest' 'southeast' 'northwest' 'northeast']


#### define numerical & categorical columns


In [21]:
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'O']

categorical_features = []

for feature in df.columns:
    if df[feature].dtype =='O':
        categorical_features.append(feature)

# print columns
print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 4 numerical features : ['age', 'bmi', 'children', 'charges']

We have 3 categorical features : ['sex', 'smoker', 'region']


In [22]:
df.head(2)

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.90,0,yes,southwest,16884.9240
1,18,male,33.77,1,no,southeast,1725.5523


### 3.8 

In [23]:
# Checking the distribution of charges based on BMI value and smoking status


file_path = 'notebook\data\medical_insurance.csv'
data = pd.read_csv(file_path)

# Group the data by smoker status and calculate the mean charges for different BMI ranges
bins = [0, 18.5, 24.9, 29.9, 39.9, 49.9, 59.9, 69.9, 79.9, 89.9, 100]
labels = ['Underweight', 'Normal weight', 'Overweight', 'Obesity I', 'Obesity II', 'Obesity III', 'Obesity IV', 'Obesity V', 'Obesity VI', 'Obesity VII']
data['bmi_category'] = pd.cut(data['bmi'], bins=bins, labels=labels, right=False)

# Group by smoker status and BMI category, then calculate the mean charges
charge_summary = data.groupby(['smoker', 'bmi_category'])['charges'].mean().unstack()
charge_summary


NameError: name 'data' is not defined